### PR TITLE
Refactor explorer service and add internal addresses fetching

### PR DESCRIPTION
**Pull request by flicker-harmony**
## Issue

Refactored explorer service to have one request GetAddresses for internal monitoring and removed deprecated code. We should make this explorer service server closed for external calls and make it available only for internal calls, like allowing only calling from localhost for example.

GetAddresses has routing /addresses. It accepts two args: size is how much addresses we want to read (default is 1000) and address prefix from which address to start reading addresses (they are stored as keys in lexicographically increasing order in eth leveldb).

### Unit Test Coverage

Before:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

After:

```
<!-- copy/paste 'go test -cover' result in the directory you made change -->
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

   NO

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

3. **Describe how the plan was tested.**

4. **How much minimum baking period after the last flag epoch should we allow on Pangaea before promotion onto mainnet?**

5. **What are the planned flag epoch numbers and their ETAs on Pangaea?**

6. **What are the planned flag epoch numbers and their ETAs on mainnet?**

    Note that this must be enough to cover baking period on Pangaea.

7. **What should node operators know about this planned change?**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    NO

9. **Does the existing `node.sh` continue to work with this change?**

10. **What should node operators know about this change?**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    NO

### CODE REVIEW

File: api/service/explorer/storage.go [hunk 1]
```diff
@@ -151,10 +100,36 @@ func (storage *Storage) UpdateAddressStorage(batch ethdb.Batch, addr string, exp
...
-		if err := batch.Put([]byte(key), encoded); err != nil {
+		if err := batch.Put([]byte(addr), encoded); err != nil {
 			utils.Logger().Warn().Err(err).Msg("cannot batch address")
 		}
 	} else {
 		utils.Logger().Error().Err(err).Msg("cannot encode address")
 	}
 }
+
+// GetAddresses ..
```
Comment (LeoHChen) [inline api/service/explorer/storage.go]: comments?

File: api/service/explorer/storage.go [hunk 2]
```diff
@@ -151,10 +100,36 @@ func (storage *Storage) UpdateAddressStorage(batch ethdb.Batch, addr string, exp
...
+	iterator := db.NewIterator()
+	if prefix != "" {
+		key := GetAddressKey(prefix)
+		iterator = db.NewIteratorWithPrefix([]byte(key))
+		if !iterator.Valid() {
+			return nil, errors.New("not valid address prefix")
+		}
+	}
+	addresses := make([]string, 0)
+	for i := 0; i < size; i++ {
```
Comment (LeoHChen) [inline api/service/explorer/storage.go]: how the pagination work here?  you need to pass in a page number, right?

**Comment by flicker-harmony** [comment]
Screenshot 2020-02-15 at 18 56 02

File: api/service/explorer/storage.go [hunk 3]
```diff
@@ -3,51 +3,34 @@ package explorer
...
 
 // Constants for storage.
 const (
-	BlockHeightKey  = "bh"
-	BlockInfoPrefix = "bi"
-	BlockPrefix     = "b"
-	TXPrefix        = "tx"
-	AddressPrefix   = "ad"
+	AddressPrefix = "ad"
+	PrefixLen     = 3
```
[the remark this answers is not in the document]
Comment (flicker-harmony) [reply api/service/explorer/storage.go]: "ad_"

File: api/service/explorer/storage.go [hunk 4]
```diff
@@ -66,68 +49,46 @@ func (storage *Storage) Init(ip, port string, remove bool) {
 	if remove {
 		var err = os.RemoveAll(dbFileName)
 		if err != nil {
-			utils.Logger().Error().Err(err).Msg("Failed to remove existing database files")
+			utils.Logger().Error().Err(err).Msg("Failed to re	move existing database files")
```
[the remark this answers is not in the document]
Comment (flicker-harmony) [reply api/service/explorer/storage.go]: Ok

**Pull request merged**